File: android/src/main/java/im/zego/zego_express_engine/ZegoMediaPlayerAudioManager.java
package im.zego.zego_express_engine;

import org.json.JSONObject;

import java.nio.ByteBuffer;

import im.zego.zegoexpress.ZegoMediaPlayer;
import im.zego.zegoexpress.callback.IZegoMediaPlayerAudioHandler;
import im.zego.zegoexpress.entity.ZegoAudioFrameParam;

public class ZegoMediaPlayerAudioManager implements IZegoMediaPlayerAudioHandler {

    private volatile static ZegoMediaPlayerAudioManager singleton;

    private static IZegoFlutterMediaPlayerAudioHandler mHander;

    public ZegoMediaPlayerAudioManager() {

    }

    /**
     * Get the custom audio capture manager instance
     */
    public static synchronized ZegoMediaPlayerAudioManager getInstance() {
        if (singleton == null) {
            singleton = new ZegoMediaPlayerAudioManager();
        }
        return singleton;
    }

    /**
     * Set audio data callback handler of the media player.
     *
     * Available since: 2.1.0
     * Description: By setting this callback, the audio data of the media resource file played by the media player can be called back.
     * When to call: After the [ZegoMediaPlayer] instance created.
     * Restrictions: None.
     * Caution: When you no longer need to get the audio frame data, please call this function again to clear the handler to stop the audio frame data callback.
     *
     * @param handler audio data callback handler for media player
     */
    public void setAudioHandler(IZegoFlutterMediaPlayerAudioHandler handler) {
        mHander = handler;
    }

    @Override
    public void onAudioFrame(ZegoMediaPlayer mediaPlayer, ByteBuffer data, int dataLength, ZegoAudioFrameParam param) {
        if (mHander != null) {
            ZGFlutterAudioFrameParam audioFrameParam = new ZGFlutterAudioFrameParam();
            audioFrameParam.channel = ZGFlutterAudioChannel.getZGFlutterAudioChannel(param.channel.value());
            audioFrameParam.sampleRate = ZGFlutterAudioSampleRate.getZGFlutterAudioSampleRate(param.sampleRate.value());
            mHander.onAudioFrame(mediaPlayer.getIndex(), data, dataLength, audioFrameParam);
        }
    }
}

File: windows/ZegoMediaPlayerAudioManager.cpp
#include "include/zego_express_engine/ZegoMediaPlayerAudioManager.h"

#include <ZegoExpressSDK.h>

static std::shared_ptr<ZegoMediaPlayerAudioManager> instance_ = nullptr;
static std::once_flag singletonFlag;

class ZegoMediaPlayerAudioHandler : public ZEGO::EXPRESS::IZegoMediaPlayerAudioHandler {
public:
    void setAudioHandler(std::shared_ptr<IZegoFlutterMediaPlayerAudioHandler> handler) {
        handler_ = handler;
    }
private:
    std::shared_ptr<IZegoFlutterMediaPlayerAudioHandler> handler_ = nullptr;

    void onAudioFrame(ZEGO::EXPRESS::IZegoMediaPlayer * mediaPlayer, const unsigned char * data,
                              unsigned int dataLength, ZEGO::EXPRESS::ZegoAudioFrameParam param) override {
        if (handler_) {
            ZGFlutterAudioFrameParam frameParam;
            frameParam.sampleRate = (ZGFlutterAudioSampleRate)param.sampleRate;
            frameParam.channel = (ZGFlutterAudioChannel)param.channel;
            handler_->onAudioFrame(mediaPlayer->getIndex(), data, dataLength, frameParam);
        }
    }
};

std::shared_ptr<ZegoMediaPlayerAudioManager> ZegoMediaPlayerAudioManager::getInstance() {
    std::call_once(singletonFlag, [&] {
        instance_ = std::make_shared<ZegoMediaPlayerAudioManager>();
    });
    return instance_;
}

std::shared_ptr<ZEGO::EXPRESS::IZegoMediaPlayerAudioHandler> ZegoMediaPlayerAudioManager::getHandler() {
    return zegoHandler_;
}

void ZegoMediaPlayerAudioManager::setAudioHandler(std::shared_ptr<IZegoFlutterMediaPlayerAudioHandler> handler) {
    handler_ = handler;
    if (handler_ && !zegoHandler_) {
        zegoHandler_ = std::make_shared<ZegoMediaPlayerAudioHandler>();
    }
    dynamic_cast<ZegoMediaPlayerAudioHandler*>(zegoHandler_.get())->setAudioHandler(handler_);
}